### DISCUSSION TITLE
Writing a "shadertoy" like program in vsg as a learning exercise

### DISCUSSION THREAD

**Discussion by dov** [vsg-dev/VulkanSceneGraph#1238] [General]
Shadertoy is a online web toy that allows the program to write a fragment shader program, while the vertex shader just passes through the four coordinates of a unit quad that fills the viewport.

The fragment shader  has a few uniforms, e.g. iTime, iTimeDelta, and iResolution.

I'd like to write a vsg based program, as a learning exercise, that replicates this behavior. But I have a few questions:

1. Is is possible to declare the above variables a pushConstants? How would you do that in VSG?
2. Alternatively how would you declare and assign them as uniforms? How do you declare "the same datastructure" in vsg and in the shader?

Thanks in advance!

**Comment by robertosfield** [comment]
The VSG provides shader compilation as long as you compile the VSG with glslang, which would make it easy for an editor to provide new GLSL code and have it compiled automatically and displayed.

I think vsgImGui would be a natural tool to use for editing shaders and controlling input values. Personally I'd allow the program to provide any combination of vertex, geometry, fragment etc. shaders, and allow models to be loaded as well as default quad across the whole window.

The most straight forward way to implement this would be to provide the data would be as uniforms.  The VSG already can provide the ViewportData as part of the vsg::ViewDependentState - the vsgPoints library's [brick.vert](https://github.com/vsg-dev/vsgPoints/blob/master/data/shaders/brick.vert) leverages this controlling point sizes.

You could also provide user defined input values as well if you so wished.

Essentially it could be shader toy that allows you to play with far more than the web browser based cousin.

**Comment by robertosfield** [comment]
The VSG is supports #pragma(tic) shader composition so you could enable optional compilation of different code paths.  If you built the functionality on top of vsg::ShaderSet then you could also output complete definitions of all the data types and functions in the ShaderSet making it possible to use with other VSG based applications.

**Comment by dov** [comment]
Thanks for your comments, @robertosfield . My initial goal is much less ambitious. I just want to be able to read a fragmant shader source file given on the command line and render it like in shadertoy.

Following your suggestion, I had a look at the `ShaderSet` as it is used e.g. in `vsgPoints::createPointsFlatShaderSet()` and while I understand the relation between the uniforms in the shader and how they are referenced by `addDescriptorBinding()` I failed to find any references of how to set the uniforms. E.g. the uniform symbol `lightData` which is used in the `brick_phong.frag` shader, is not referenced anywhere outside of `BrickShaderSet.cpp`. So how can you change the uniform value from c++? And how do you trigger a rewdraw with the new values after you change it?

**Reply by robertosfield** [reply]
The discussion #1211 has details of updating data dynamically.

With vsg::ShaderSet, it's essentially an implementation of reflection/introspection for sets of shaders that are written to work together, with the options encoded as #defines in the shaders being listed along with any uniform, textures or vertex attributes they are associated with, along with the types of those uniform, textures or vertex attributes.  

The ShaderSet provides the high level description of the collection of shaders and associated state in a way that tools like vsg::Builder or the vsgXchange::assimp can use along with vsg::GraphicsPipelineConfigurator to build the required StateGroup/BindGraphicsPipeline/BindDescriptorSet etc. through to the layout/data types of the vertex attributes that are assigned to vsg::Draw/DrawIndex/VertexIndexDraw/Geometry classes.

You needn't use ShaderSet you can just rolled your own scene graph as many of the VSG examples do, so for you are a developer that's new to VSG/Vulkan it may help you to learn the lower level classes like BindGraphicsPipeline/GraphicsPipeline/ShaderModule etc.

As time goes on you'll hopefully master both the lower and higher levels and know how they all relate to each.  With a shadertoy example you'll likely get others helping out as well, as it's an interesting and fun exercise.
